dashboard/test/controllers/lti_provider_controller_test.rb

```diff
@@ -0,0 +1,206 @@
+
+    # LTI does not typically send gender
+    assert_equal nil, user.gender
+  end
+
+  test "LTI sign-in with new user_id can also create new Teacher account" do
+    params = lti_consumer_params(
+      TEST_CONSUMER_KEY,
+      TEST_PROVIDER_KEY, method_name,
+      "user_id" => "12345",
+      "lis_person_name_full" => "Cat Hat",
+      "roles" => "Instructor")
+
+    assert_creates(User) do
+      lti_post :sso, params
+    end
+
+    # Teachers are defined to be 21 years old, emails are stored
+    user = User.last
+    assert_equal 'Cat Hat', user.name
+    assert_equal "12345", user.uid
+    assert_equal User::TYPE_TEACHER, user.user_type
+    assert_equal "21+", user.age
+  end
+
+  test "LTI sign-in can also handle multiple LTI roles" do
+    params = lti_consumer_params(
+      TEST_CONSUMER_KEY,
+      TEST_PROVIDER_KEY,  method_name,
+      "user_id" => "12345",
+      "lis_person_name_full" => "Cat Hat",
+      "roles" => "Instructor, TeachingAssistant, ContentDeveloper")
+
+    assert_creates(User) do
+      lti_post :sso, params
+    end
+
+    user = User.last
+    assert_equal "Cat Hat", user.name
+    assert_equal "12345", user.uid
+    assert_equal User::TYPE_TEACHER, user.user_type
+  end
+
+  test "LTI sign-in can also handle multiple Learner roles (edge case)" do
+    params = lti_consumer_params(
+      TEST_CONSUMER_KEY,
+      TEST_PROVIDER_KEY,  method_name,
+      "user_id" => "12345",
+      "lis_person_name_full" => "Cat Hat",
+      "custom_age" => 9,
+      "roles" => "Learner, Learner")
+
+    assert_creates(User) do
+      lti_post :sso, params
+    end
+
+    user = User.last
+    assert_equal 'Cat Hat', user.name
+    assert_equal "12345", user.uid
+    assert_equal User::TYPE_STUDENT, user.user_type
+    assert_equal 9, user.age
+  end
+
+  test "LTI sign-in creates new User with no password and provider set" do
+    params = lti_consumer_params(
+      TEST_CONSUMER_KEY,
+      TEST_PROVIDER_KEY,  method_name,
+      "user_id" => "12345",
+      "lis_person_name_full" => "Cat Hat",
+      "custom_age" => 8,
+      "roles" => "Learner")
+
+    assert_creates(User) do
+      lti_post :sso, params
+    end
+
+    user = User.last
+    assert_equal "lti_#{TEST_CONSUMER_KEY}", user.provider
+    assert_equal "12345", user.uid
+    assert_nil user.password
+  end
+
+  test "LTI sign-in with no user_id does not sign in" do
+    params = lti_consumer_params(
+      TEST_CONSUMER_KEY,
+      TEST_PROVIDER_KEY, method_name,
+      "custom_age" => 8,
+      "lis_person_name_full" => "Cat Hat")
+
+    assert_does_not_create(User) do
+      lti_post :sso, params
+    end
+
+    assert_redirected_to 'http://test.host/users/sign_up'
+    assert_match "[user_id] missing", session['flash'].to_s
+  end
+
+  test "LTI sign-in with bad oauth_timestamp does not sign in" do
+    params = lti_consumer_params(
+      TEST_CONSUMER_KEY,
+      TEST_PROVIDER_KEY,  method_name,
+      "oauth_timestamp" => Time.now.to_i - 6.minutes,
```

Comment:
minor: Rails best practice is to use `Time.zone.now` instead of `Time.now`. See http://api.rubyonrails.org/classes/ActiveSupport/TimeWithZone.html. This doesn't matter on production however, because the time zone on those servers is utc, so the two are equivalent.

Why - 6.minutes?